Add Guide: Read Profile Data (#112)
lukso-network/docs: docs/guides/fetch-data/01-read-profile-data.md

```javascript title="read_profile.js"
// Import and Network Setup
const Web3 = require('web3');
const web3 = new Web3('https://rpc.l14.lukso.network');

// Our static Universal Profile address
const sampleProfileAddress = '0x0C03fBa782b07bCf810DEb3b7f0595024A444F4e';

/*
 * Check if @param is a valid blockchain address
 * @returns: boolean
 */
function isAddrValid(address) {
  let formattedAddress = web3.utils.toChecksumAddress(address);
  return web3.utils.checkAddressChecksum(formattedAddress);
}

// Debug
console.log(isAddrValid(sampleProfileAddress));
```

If the function `isAddrValid` gives us back `true`, the address is valid and can be checked for contact data.

## Step 2 - Call the Universal Profile

To inspect the address and check if it has an ERC725 contract, we can call its interface through the `erc725.js` library. The instance of the contract will need the following information:

- Schema which describes contract functions
  - SupportedStandards will fetch the interface
  - LSP3Profile fetches the data of the profile
  - LSP1UniversalReceiverDelegate will fetch received assets

> - Provider has the RPC address of the called network
> - Address has the related blockchain address of the contract
> - Config has the base URL of the IPFS server

Reviewer: ```suggestion
- `address`: the address of the contract
- `provider`: a [provider](https://docs.lukso.tech/tools/erc725js/providers) object. Usually used with the RPC endpoint URL
- `config`:  used to configure the IPFS gateway
```

go straight to the point - the more text, the less people will read.

> Besides the schema, we also use `isomorphic-fetch` to fetch the HTTP response from the profile while using `node` for execution. You may not need this library if you use extended environments like `ReactJS` or `VueJS`. First, we have to install the libraries.

Reviewer: ```suggestion
Besides the schema, we also use `isomorphic-fetch` to fetch the HTTP response from the profile while using `node` for execution. You may not need this library if you use browser environments like `ReactJS` or `VueJS`. First, we have to install the libraries.
```

```shell

> npm install @erc725/erc725.js
> npm install isomorphic-fetch

Reviewer: ```suggestion
npm install @erc725/erc725.js isomorphic-fetch
```

```

> Within the JaveScript file, we can add this code snippet right underneath. After we embed the ERC725 object and the fetch functionality, we can declare all data needed to instantiate the ERC725 contract instance. As discussed above, our schema, provider, and network configuration need to be declared. While fetching, we can use the function from the previous step to handle errors that might occur.

Reviewer: ```suggestion
After we embed the ERC725 object and the fetch functionality, we can declare all data needed to instantiate the ERC725 contract instance. While fetching, we can use the function from the previous step to handle errors that might occur.
```

"we embed" ? You mean "we install" ?

ERC725 -> `erc725.js`

Author: I will use the term "import"

    valueType: 'bytes',
  },
  {
    name: 'LSP3Profile',
    key: '0x5ef83ad9559033e6e941db7d7c495acdce616347d28e90c7ce47cbfcfcad3bc5',
    keyType: 'Singleton',
    valueContent: 'JSONURL',
    valueType: 'bytes',
  },
  {
    name: 'LSP1UniversalReceiverDelegate',
    key: '0x0cfc51aec37c55a4d0b1a65c6255c4bf2fbdf6277f3cc0730c45b828b6db8b47',
    keyType: 'Singleton',
    valueContent: 'Address',
    valueType: 'address',
  },
];

const provider = new Web3.providers.HttpProvider(
  'https://rpc.l14.lukso.network',
);

const config = {
  ipfsGateway: 'https://ipfs.lukso.network/ipfs/',
};

/*
 * Try fetching the @param's Universal Provile data
 * @returns: string JSON or custom error
 */
async function getProfile(address) {
  if (isAddrValid(address)) {
    try {
      const erc725 = new ERC725(erc725schema, address, provider, config);
      return await erc725.fetchData();
    } catch (err) {
      return console.log('This is not an ERC725 Contract');
    }
  } else {
    return console.log('This is not an blockchain address');
  }
}

// Debug
getProfile(sampleProfileAddress).then((profileData) =>
  console.log(JSON.stringify(profileData, undefined, 2)),
);
```

> If everything went fine, we now have the profile's dataset as a universal JSON structure that looks like this:

Reviewer: ```suggestion
If everything went fine, we now have the profile's [LSP3 - Universal Profile Metadata](https://docs.lukso.tech/standards/universal-profile/lsp3-universal-profile-metadata) JSON. It should look like this:
```

⚠️ -> use relative link not absolute one. like: `../../xxxx`

<details>

>     <summary>Show JSON responce</summary>

Reviewer: ```suggestion
    <summary>Show JSON response</summary>
```

```json
{
  "SupportedStandards:ERC725Account": "0xafdeb5d6",
  "LSP3Profile": {
    "LSP3Profile": {
      "name": "...",
      "links": [
        {
          "title": "... ",
          "url": "..."
        },
        ...
      ],
      "description": "...",
      "profileImage": [
        {
          "width": 1512,
          "height": 1998,
          "hashFunction": "keccak256(bytes)",
          "hash": "0x...",
          "url": "..."
        },
        ...
      ],
      "backgroundImage": [
        {
          "width": 1512,
          "height": 1998,
          "hashFunction": "keccak256(bytes)",
          "hash": "0x...",
          "url": "..."
        },
        ...
      ],
      "tags": ["sample profile"]
    }
  },
  "LSP1UniversalReceiverDelegate": "0x..."
}
```

</details>

## Step 3 - Read the Dataset

With the JSON response, we can fetch all sorts of data:

- Profile metadata
- Pictures and their properties
- Receiver address containing asset data

> Just instantiate variables for the information you would like to process. All demo fetch functions provided with this guide will come with a debug parameter. You can set it to `1` to see results directly in your terminal.

Reviewer: i'm not sure if it is useful to add code and logic for this debug thing.

Just keep it super simple - add the console.log in your code.

If people want to remove them it is up to them.

Put the focus on things that matter. Having a debug argument does not matter and spread the focus on non important things.

<details>
    <summary>Fetch the profile's metadata</summary>

```javascript title="read_profile.js"
...
// Fetchable metadata information
let name;
let description;

let links = [];
let firstLinkTitle;
let firstLinkURL;

let tags = [];
let firstTag;

/*
 * Fetch metadata information from the JSON dataset of
 * an Universal Profile
 */
async function fetchProfileData(enableDebug) {
  profileData = await getProfile(sampleProfileAddress);
  name = profileData.LSP3Profile.LSP3Profile.name;
  description = profileData.LSP3Profile.LSP3Profile.description;

  links = profileData.LSP3Profile.LSP3Profile.links;
  firstLinkTitle = links[0].title;
  firstLinkURL = links[0].url;

  tags = profileData.LSP3Profile.LSP3Profile.tags;
  firstTag = tags[0];

>   if (enableDebug === 1) {

Reviewer: cf comment above - just output it. dont deal with this debug -

    console.log('Name ' + name);
    console.log('Description: ' + description + '\n');
    console.log('Links: ' + JSON.stringify(links, undefined, 2) + '\n');
    console.log('Title of first Link: ' + firstLinkTitle);
    console.log('URL of first Link: ' + firstLinkURL + '\n');
    console.log('Tags: ' + JSON.stringify(tags, undefined, 2) + '\n');
    console.log('First tag: ' + firstTag + '\n');
  }
}

// Debug
fetchProfileData(1);
```

</details>

<details>
    <summary>Fetch the profile's picture properties</summary>

```javascript title="read_profile.js"
...
// Fetchable picture information
let baseURL = 'https://ipfs.lukso.network/ipfs/';

let backgroundImageLinks = [];
let fullSizeBackgroundImg;

let profileImageLinks = [];
let fullSizeProfileImg;

/*
 * Fetch picture information from the JSON dataset of

>  * an Universal Profile

Reviewer: ```suggestion
 * a Universal Profile
```

 */
async function fetchPictureData(enableDebug) {
  pictureData = await getProfile(sampleProfileAddress);
  let backgroundImagesIPFS =
    pictureData.LSP3Profile.LSP3Profile.backgroundImage;
  let profileImagesIPFS = pictureData.LSP3Profile.LSP3Profile.profileImage;

  for (let i in backgroundImagesIPFS) {
    backgroundImageLinks.push([
      i,
      baseURL + backgroundImagesIPFS[i].url.substring(7),
    ]);
  }

  for (let i in profileImagesIPFS) {
    profileImageLinks.push([
      i,
      baseURL + profileImagesIPFS[i].url.substring(7),
    ]);
  }

>   fullSizeBackgroundImg = backgroundImageLinks[0][1];

Reviewer: to do it correctly, you need to try/catch or handle case where there is no images.
The JSON documents may not be valid as they are set by anyone

---
sidebar_label: 'Read Profile Data'
sidebar_position: 3.1
---

# Read Universal Profile Data

In this guide, we will learn how to:

- call a Universal Provile address
- read data from a Universal Profile

![Universal Profile example on universalprofile.cloud](./img/example-up.png)

We will use:

- [web3.js](https://web3js.readthedocs.io/en/v1.7.0/) for utility as well as connecting with the LUKSO network

> - [erc725.js](https://docs.lukso.tech/tools/erc725js/getting-started/) library to check the interface of an profile.

Reviewer: can you use a relative link (`../../`) instead of using the public URL.

This allows us to keep track of broken links and also make this more robust

```javascript title="read_profile.js"
// Import and Network Setup
const Web3 = require('web3');
const web3 = new Web3('https://rpc.l14.lukso.network');

// Our static Universal Profile address
const sampleProfileAddress = '0x0C03fBa782b07bCf810DEb3b7f0595024A444F4e';

/*
 * Check if @param is a valid blockchain address
 *
 * @returns: boolean
 */
function isValidAddress(address) {
  let formattedAddress = web3.utils.toChecksumAddress(address);
  return web3.utils.checkAddressChecksum(formattedAddress);
}

// Debug
console.log(isValidAddress(sampleProfileAddress));
```

If the function `isValidAddress` gives us back `true`, the address is valid and can be used.`

## Step 2 - Call the Universal Profile

To inspect the address and check if it has an ERC725 contract, we can call its interface through the `erc725.js` library. The instance of the contract will need the following information:

> -[LSP2 - ERC725Y JSON Schema](--add link to this page: https://docs.lukso.tech/standards/generic-standards/lsp2-json-schema/) which describes contract functions

Reviewer: -> please update the link -> `../../`

the suggestion was not fully ready.

- SupportedStandards will fetch the interface
- LSP3Profile fetches the data of the profile
- LSP1UniversalReceiverDelegate will fetch received assets
- `address`: the address of the contract

> - `provider`: a [provider](https://docs.lukso.tech/tools/erc725js/providers) object. Usually used with the RPC endpoint URL

Reviewer: use relative link

    valueType: 'bytes',
  },
  {
    name: 'LSP3Profile',
    key: '0x5ef83ad9559033e6e941db7d7c495acdce616347d28e90c7ce47cbfcfcad3bc5',
    keyType: 'Singleton',
    valueContent: 'JSONURL',
    valueType: 'bytes',
  },
  {
    name: 'LSP1UniversalReceiverDelegate',
    key: '0x0cfc51aec37c55a4d0b1a65c6255c4bf2fbdf6277f3cc0730c45b828b6db8b47',
    keyType: 'Singleton',
    valueContent: 'Address',
    valueType: 'address',
  },
];

const provider = new Web3.providers.HttpProvider(
  'https://rpc.l14.lukso.network',
);

const config = {
  ipfsGateway: 'https://ipfs.lukso.network/ipfs/',
};

/*
 * Try fetching the @param's Universal Provile data
 * @returns: string JSON or custom error
 */
async function getProfile(address) {
  if (isValidAddress(address)) {
    try {
      const erc725 = new ERC725(erc725schema, address, provider, config);
      return await erc725.fetchData();
    } catch (error) {
      return console.log('This is not an ERC725 Contract');
    }
  } else {
    return console.log('This is not an blockchain address');
  }
}

// Debug
getProfile(sampleProfileAddress).then((profileData) =>
  console.log(JSON.stringify(profileData, undefined, 2)),
);
```

> If everything went fine, we now have the profile's [LSP3 - Universal Profile Metadata](https://docs.lukso.tech/standards/universal-profile/lsp3-universal-profile-metadata) JSON. It should look like this:

Reviewer: use relative link

---
sidebar_label: 'Read Profile Data'
sidebar_position: 3.1
---

# Read Universal Profile Data

In this guide, we will learn how to:

- call a Universal Provile address
- read data from a Universal Profile

![Universal Profile example on universalprofile.cloud](./img/example-up.png)

We will use:

> - [web3.js](https://web3js.readthedocs.io/en/v1.7.0/) for utility as well as connecting with the LUKSO network

Reviewer 2: Maybe be a bit more specific and says LUKSO **L14** network?

> - [erc725.js](../../tools/erc725js/getting-started/) library to check the interface of an profile.

Reviewer 2: "of **a** profile"

> - [isomorphic-fetch](https://github.com/matthew-andrews/isomorphic-fetch) to enable you to use fetch in plain Node code

Reviewer 2: suggested edit:
"`fetch()` in Node.js code"

> ## Table of Contents

Reviewer 2: Maybe this table of content is not necessary, since it appears in the sidebar?

Otherwise, if you want to keep them, I would recommend adding the internal anchor links, like so in Markdown:

```markdown
1. [Check for valid Blockchain Address](#step-1---check-for-valid-blockchain-address)

```

1. Check for valid Blockchain Address
2. Call the Universal Profile
3. Read the Dataset
   1. Fetch the profile's metadata
   2. Fetch the profile's picture properties
   3. Fetch the profile's universal receiver

## Step 1 - Check for valid Blockchain Address

> When receiving data from a Universal Profile within an application, it is commonly used as input. You want to make sure that the address is valid in the first place. You can do a checkup by converting the value into a checksum address from the `web3.js` library. We need to install the library first. You can do this by opening up the terminal in the project's folder of your choice.

Reviewer 2: Probably stick to "We", since you started like this in the beginning.

_Suggested changes:_
"When receiving data from a Universal Profile within an application, it is commonly used as input. We want to make sure that the address is valid in the first place. We can do a checkup by converting the value into a checksum address using the `web3.js` library. 

First, open a terminal in the project's folder of your choice and install the `web3.js` library."

```shell
npm install web3
```

> After installation, we can move on with the JavaScript file. During the process, we will always use the same file. To make the guide more understandable, we use a sample profile address. You will most likely exchange this static variable with a dynamic value from an input field or fetching process within your app.

Reviewer 2: "**change**", not ~exchange~

```javascript title="read_profile.js"
// Import and Network Setup
const Web3 = require('web3');
const web3 = new Web3('https://rpc.l14.lukso.network');

// Our static Universal Profile address
const sampleProfileAddress = '0x0C03fBa782b07bCf810DEb3b7f0595024A444F4e';

/*
 * Check if @param is a valid blockchain address
 *
 * @returns: boolean
 */
function isValidAddress(address) {
  let formattedAddress = web3.utils.toChecksumAddress(address);
  return web3.utils.checkAddressChecksum(formattedAddress);
}

// Debug
console.log(isValidAddress(sampleProfileAddress));
```

> If the function `isValidAddress` gives us back `true`, the address is valid and can be used.`

Reviewer 2: Adding parentheses to emphasize it is a function? Like below

`isValidAddress(...)`

## Step 2 - Call the Universal Profile

To inspect the address and check if it has an ERC725 contract, we can call its interface through the `erc725.js` library. The instance of the contract will need the following information:

> - [LSP2 - ERC725Y JSON Schema](../../standards/generic-standards/lsp2-json-schema/) which describes contract functions

Reviewer 2: @fhildeb 
This is a big error and really needs to be corrected.

LSP2 - ERC725Y JSON Schema ~**does not describe the contract functions**~. :x:
It describes the data / metadata already present in the contract storage, and which keys to use to retrieve this data. :white_check_mark:

It is ERC165 interface ID that describes the functions of a contract.

>   - SupportedStandards will fetch the interface

Reviewer 2: This is incorrect.

You have also have to give the Metadata Standard to check for, so for instance in this guide, the key would be `SupportedStandards:LSP3UniversalProfile`
